src/Microsoft.AspNetCore.Mvc.Core/Internal/ActionDescriptorCollectionProvider.cs

```diff
@@ -35,36 +49,48 @@ public ActionDescriptorCollection ActionDescriptors
         {
             get
             {
-                if (_collection == null)
+                var result = _collection;
+
+                if (result == null)
                 {
-                    _collection = GetCollection();
+                    result = GetCollection();
+                    _collection = result;
```

Comment:
We might end up overwriting a newer version of the collection with an older one here. Is this problematic?